tinygrad/codegen/linearizer.py

```diff
@@ -21,25 +21,86 @@ class UOps(Enum):
+  if valid.min == 0 and isinstance(idx, ModNode) and isinstance(idx.a, SumNode):
+    nodes = valid.nodes if isinstance(valid, AndNode) else [valid]
+    same_dict: Dict[Node, List[Tuple[int, Node]]] = {}
+    idx_nodes = idx.a.flat_components
+
+    for node in nodes:
+      if not isinstance(node, LtNode) or not isinstance(node.a, SumNode): continue
+
+      nd_flat, nd_vars = node.a.flat_components, node.vars()
+
+      same = [x for x in idx_nodes if (x.a if isinstance(x, MulNode) else x) in nd_vars]
+
+      if len(same) != len(nd_vars): continue
+
+      first_b, second_b = nd_flat[0].b if isinstance(nd_flat[0], MulNode) else 1, same[0].b if isinstance(same[0], MulNode) else 1
+      k, same_sum = second_b//first_b, Variable.sum(same)
+
+      if k*(node.a) == same_sum: same_dict[same_sum] = same_dict.get(same_sum, []) + [(k, node)]
+
+    for key in same_dict.keys():
+      same, mnn, mxn = key.flat_components, key.min, key.max # type: ignore # Same is sumnode because node.a is SumNode
+      for k, node in same_dict[key]: # TODO: This part may need more thinking
+        if k < 0: mnn = (-k)*max((-node.b) + 1, min([-lal.b if isinstance(lal, MulNode) else 1 for lal in same]))
+        else: mxn = (node.b - 1)*k
+
+      fake_var = Variable("valid_fake", mnn, mxn)
+      total = (Variable.sum([x for x in idx_nodes if x not in same]) + fake_var) % idx.b
+      idx = total.substitute({fake_var: key})
+      # TODO: If idx has no ModNode we may can remove the valid node, but removing it needs careful thinking
+
+  # Simplify SumNodes
+  # This part just removes valid nodes if node is exactly same as idx or idy
+  # idx = 3*a + b (+ 5), valid = 3*a + b < 10 # Valid will be removed as idx will go out of bounds
+  # Check for var intersection, removing valid can affect other index
+  if valid.min == 0 and not idx_vars.intersection(idy_vars):
+    nds = valid.nodes if isinstance(valid, AndNode) else [valid]
+    flats = [id.flat_components for id in (idx, idy) if isinstance(id, SumNode)]
+    sym_sums = [Variable.sum([i for i in flat if not isinstance(i, NumNode)]) for flat in flats]
+    ones = [node for sym_sum in sym_sums for node in nds if (node.a == sym_sum) or (-(node.a) == sym_sum)] # type: ignore # AndNode always consists of LtNode
+    valid = Variable.ands([i for i in nds if i not in ones])
+
+  # This is the slow part
+  # This part is for brute forcing all possible values of idx, idy and valid
+  # If valid is both 0 and 1 for the same (idx, idy) we can not delete the valid
+  if valid.min == 0 and not isinstance(idx, ModNode):
+    variables = tuple(val_vars | idy_vars | idx_vars)
+    val_infer, idx_infer, idy_infer = valid.expand(variables), idx.expand(variables), idy.expand(variables)
+    val_dict: Dict[int, Set[Tuple[int,int]]] = {0:set(), 1:set()}
+
+    for v, x, y in zip(val_infer, idx_infer, idy_infer): val_dict[v.min].add((x.min, y.min))
+
+    if not val_dict[1].intersection(val_dict[0]): valid = NumNode(1)
```

Comment:
This might be faster. It breaks early if a pair is found. 
```suggestion
    val_list: List[Set[Tuple[int,int]]] = [set(), set()]
    
    for v, x, y in zip(valid.expand(variables), idx.expand(variables), idy.expand(variables)): 
      if (x.min, y.min) in val_dict[1-v.min]: break # found in other set
      val_dict[v.min].add((x.min, y.min))
    else:
      valid = NumNode(1) # no pair found
```

Reply:
This is I believe faster but does it make difference as main slow part is expands.

Also, I think intersection is more easier to understand but I will look at this ones oerformance as well